staging/src/k8s.io/dynamic-resource-allocation/utils/pci.go

```diff
@@ -0,0 +1,213 @@
+package utils
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+var (
+	// bdfRegexp matches PCI address in BDF notation.
+	// The format is <domain>:<bus>:<device>.<function>
+	// where:
+	// - domain: 4-digit hexadecimal number representing the PCI domain
+	// - bus: 2-digit hexadecimal number representing the PCI bus
+	// - device: 2-digit hexadecimal number representing the PCI device
+	// - function: 1-digit hexadecimal number representing the PCI function
+	// Example: "0000:0e:1f.0" represents domain 0, bus 14, device 31, function 0.
+	// ref: https://wiki.xenproject.org/wiki/Bus:Device.Function_(BDF)_Notation
+	bdfRegexp = regexp.MustCompile(`^([0-9a-f]{4}):([0-9a-f]{2}):([0-9a-f]{2})\.([0-9a-f]{1})$`)
+)
+
+// PCIRoot represents a PCI root address in the combination of Domain and Bus.
+// This represents the top-level element in sysfs PCI device hierarchy.
+// (i.e. /sys/devices/pci<domain>:<bus>)
+//
+// ref: https://docs.kernel.org/PCI/sysfs-pci.html
+type PCIRoot struct {
+	Domain uint32
+	Bus    uint32
+}
+
+func (p *PCIRoot) String() string {
+	return fmt.Sprintf("%04x:%02x", p.Domain, p.Bus)
+}
+
+// The PCIAddress holds PCI address components in BDF notation.
+// <domain>:<bus>:<device>.<function>
+// where:
+// - domain: 4-digit hexadecimal number representing the PCI domain
+// - bus: 2-digit hexadecimal number representing the PCI bus
+// - device: 2-digit hexadecimal number representing the PCI device
+// - function: 1-digit hexadecimal number representing the PCI function
+//
+// Example: "0000:0e:1f.0" represents domain 0, bus 14, device 31, function 0.
+//
+// ref: https://wiki.xenproject.org/wiki/Bus:Device.Function_(BDF)_Notation
+type PCIAddress struct {
```

Comment:
Should we ensure that address matches and does not exceed the size limits of the spec? I.e function can only be up to 15

Reply:
Thanks for the info! Agreed. We should validate them. Would you know a pointer to official(-ish) spec? I'd like to know the specs for other parts(domain, bus, device)